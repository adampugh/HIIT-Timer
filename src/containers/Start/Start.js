import React, { Component } from "react";

import Navbar from "../../components/UI/navbarLoggedIn";
import StartWorkoutBanner from "../Start/startWorkoutBanner";
import Footer from "../../components/UI/footer";

// a for loop that calls the start timer function for each workout item
// inside the start timer function increment the count at the end of the setTimeout
// retrieve the amount in secs from the state 
// manage the secs in the function locally - no need to manipulate state as it only
// goes down in time. 
// work out total time locally as well.

// mapstatetoprops using redux to access db

// class Start extends Component {
//     render() {
//         return (
//             <div>
//                 <Navbar />
//                 <StartWorkoutBanner workout={this.props.location.state.workout} />
//                 <Footer />
//             </div>
//         );
//     }
// };

const Start = (props) => (
    <div>
        <Navbar />
        <StartWorkoutBanner workout={props.location.state.workout} />
        <Footer />
    </div>
);



export default Start;