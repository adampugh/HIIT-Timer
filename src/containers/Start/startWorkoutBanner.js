import React, { Component } from "react";
import { Redirect } from "react-router-dom";
var moment = require("moment");
var momentDurationFormatSetup = require("moment-duration-format");


class startWorkoutBanner extends Component {
    state = {
        position: 0,
        paused: true,
        totalTime: 0,
        currentName: "Start",
        currentTime: 0
    }
    
    // startTimer - get time from workout / if 0 increase position / add to total time
    startTimer = () => {
        if (this.state.currentTime === 0) {
            this.setState({
                currentTime: this.props.workout[this.state.position].time,
            });
        }
        this.setState({
            currentName: this.props.workout[this.state.position].name,
            paused: false
        });
        this.startExercise();
    }

    // display number of seconds left
    startExercise(exerciseTime) {
        this.exerciseInterval();
    }

    nextExercise() {
        this.setState({
            position: this.state.position + 1
        });
        if (this.props.workout[this.state.position]) {
            this.startTimer();
        } else {
            this.setState({
                position: 0
            })
        }
    }
    
    exerciseInterval() { 
        const interval = setInterval(() => {
            if (this.state.currentTime > 0 && !this.state.paused) {
                this.setState({
                    currentTime: this.state.currentTime - 1,
                    totalTime: this.state.totalTime + 1
                });
            } else if (this.state.paused) {
                clearInterval(interval);
            } else {
                clearInterval(interval);
                this.nextExercise();
            }
        }, 1000);
    }

    stopTimer() {
        this.setState({
            paused: true
        });
    }

    handleBannerClick = () => {
        if (this.state.paused === true) {
            this.startTimer();
        } else {
            this.stopTimer();
        }
    }

    // format seconds as minutes in the render function

    render() {
        let customColor;
        if (this.props.workout[this.state.position]) {
            customColor = {
                background: this.props.workout[this.state.position].color
            }
        }
        
        let startWorkoutBannerPage = this.props.workout.length < 1 ? (
            <Redirect to="/workouts" />
        ) : (
            <div className="startWorkoutBanner" onClick={this.handleBannerClick} style={customColor}>
                <h1>{this.state.currentName}</h1>
                <h1>{moment.duration(this.state.currentTime, "seconds").format("mm:ss")}</h1>
                <div className="startWorkoutBanner__grid">
                    <div>
                        <h3>Position</h3>
                        <h2>{this.state.position + 1} / {this.props.workout.length}</h2>
                    </div>
                    <div>
                        <h3>{this.state.paused ? "Play" : "Pause"}</h3>
                        <h3>{this.state.paused ? '▶' : "▌▌"}</h3>
                    </div>
                    <div>
                        <h3>Total Time</h3>
                        <h2>{moment.duration(this.state.totalTime, "seconds").format("mm:ss")}</h2>
                    </div>
                </div>
            </div>
        )
        return startWorkoutBannerPage;
    }
};



// '▷' : '❘ ❘'

// <h2>{this.state.paused ? <span><i className="fas fa-play"></i></span>
//     : <span><i className="fas fa-pause"></i></span> }</h2>


export default startWorkoutBanner;