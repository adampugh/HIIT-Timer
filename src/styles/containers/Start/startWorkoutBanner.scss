.startWorkoutBanner {
    cursor: pointer;
    text-align: center;
    color: white;
    background: $purple-gradient;
    height: 90vh;
    -webkit-background-size: cover;
    -moz-background-size: cover;
    -o-background-size: cover;
    background-size: cover;
    padding: 60px;
    h1 {
        margin: 0;
        padding: 40px 0;
        font-size: 4em;
    }
}

.startWorkoutBanner__grid {
    display: grid;
    grid-template-columns: 1fr 1fr 1fr;
    
    padding: 40px 0;
    div {
        border: 1px solid white;    
    }
}

@media only screen and (min-width: $desktop-breakpoint) {
    .startWorkoutBanner {
       height: 80vh;
    }
}