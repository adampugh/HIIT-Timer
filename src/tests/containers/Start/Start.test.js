import React from "react";
import { shallow } from "enzyme"; 

import Start from "../../../containers/Start/Start";
import workoutState from "../../fixtures/workouts";

test("should render Start correctly", () => {
    let locationObj = {
        state: {
            workout: workoutState.workouts[0].exercises,
            id: workoutState.workouts[0].id
        }
    };
    const wrapper = shallow(<Start location={locationObj} />);
    expect(wrapper).toMatchSnapshot();
});