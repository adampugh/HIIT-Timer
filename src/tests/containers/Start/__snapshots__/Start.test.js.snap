// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Start correctly 1`] = `
<div>
  <navbarLoggedIn />
  <startWorkoutBanner
    workout={
      Array [
        Object {
          "color": "#40b3e1",
          "id": "L2453",
          "name": "squats",
          "time": 60,
        },
        Object {
          "color": "#40b3e1",
          "id": "L2435y",
          "name": "pressups",
          "time": 60,
        },
      ]
    }
  />
  <footer />
</div>
`;
